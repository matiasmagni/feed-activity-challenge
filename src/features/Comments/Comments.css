label {
    display: inline-flex;
}

label * {
    margin-left: 5px;
    width: 620px;
    margin-bottom: 20px;
}

.input-name {
    width: 599px;
}

.input-email {
    width: 602px;
}

.comments {
    width: 570px;
    height: 120px;
}

.submit-container {
    text-align: right;
    margin-bottom: 10px;
}
