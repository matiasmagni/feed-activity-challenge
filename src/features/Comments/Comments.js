import React, { useState, useEffect } from 'react';
import { useParams } from "react-router-dom";
import { useSelector, useDispatch } from 'react-redux';
import axios from 'axios';
import config from '../../config';
import { selectComments, addComment } from '../../redux/commentsSlice';
import './Comments.css';

const Comments = () => {
    let { postId } = useParams();
    const dispatch = useDispatch();
    const [post, setPost] = useState(null);
    const [comments, setComments] = useState([]);
    const [nameInputValue, setNameInputValue] = useState('');
    const [emailInputValue, setEmailInputValue] = useState('');
    const [commentTextValue, setCommentTextValue] = useState('');

    useEffect(() => {
        const fetchPost = async (postId) => {
            const result = await axios(`${config.API_URL.POSTS}/${postId}`);
            setPost(result.data);
        };

        const fetchComments = async (postId) => {
            const result = await axios(`${config.API_URL.COMMENTS}?postId=${postId}`);
            //dispatch(setComments({ comments: result.data }));
            setComments(result.data);
        };

        fetchPost(postId);
        fetchComments(postId);
    }, [postId]);

    const changeName = (event) => {
        setNameInputValue(event.target.value);
    }

    const changeEmail = (event) => {
        setEmailInputValue(event.target.value);
    }

    const changeComment = (event) => {
        setCommentTextValue(event.target.value);
    }

    const handleSubmit = (event) => {
        event.preventDefault();

        const newComment = {
            postId: postId,
            name: nameInputValue,
            email: emailInputValue,
            body: commentTextValue,
        };

        console.log()

        dispatch(addComment(newComment));
        setComments([newComment, ...comments]);
    }

    return (
        <div className="container-fluid video-player">
            <div className="container">
                <div className="row vr-gallery">
                    {post &&
                        <>
                            <div id={post.id} className="col-md-8 mb-4">
                                <div className="row">
                                    <div className="col-md-12 col-lg-7 pr-0 pd-md">
                                        <img src="/images/gallery-img1.png" alt="" />
                                    </div>
                                    <div className="col-md-12 col-lg-5 light-bg cus-pd cus-arrow-left">
                                        <h3>{post.title}</h3>
                                        <p>{post.body}</p>
                                    </div>
                                </div>
                            </div>
                            <div id="user-comment" className="col-md-8 mb-4">
                                <div className="row">
                                    <div className="col-md-12 col-lg-12 light-bg cus-pd cus-arrow-left">
                                        <h3>Leave a comment here:</h3>
                                        <form onSubmit={handleSubmit}>
                                            <label>
                                                Name:
                                                <input id="name" className="input-name" name="name" type="text" value={nameInputValue} onChange={changeName} />
                                            </label>
                                            <label>
                                                Email:
                                                <input id="email" className="input-email" name="email" type="email" value={emailInputValue} onChange={changeEmail} />
                                            </label>
                                            <label>
                                                Comment:
                                                <textarea className="comments" value={commentTextValue} onChange={changeComment} />
                                            </label>
                                            <div className="submit-container">
                                                <input id="addComment" name="addComment" type="submit" value="Publish" />
                                            </div>
                                        </form>
                                    </div>
                                </div>
                            </div>
                        </>

                    }
                    {comments &&
                        <div id="comments" className="col-md-8 mb-4">
                            {comments.map(comment => (
                                <div key={`comment-${comment.id}`} className="row">
                                    <div className="col-md-12 col-lg-12 light-bg cus-pd cus-arrow-left">
                                        <h3>{comment.name}</h3>
                                        <h6>{comment.email}</h6>
                                        <p>{comment.body}</p>
                                    </div>
                                </div>
                            ))}
                        </div>
                    }
                </div>
            </div>
        </div>
    );
};

export default Comments;
