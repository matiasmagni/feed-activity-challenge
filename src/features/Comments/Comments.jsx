import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import axios from 'axios';
import config from '../../config';
import { selectComments, addComment } from '../../redux/commentsSlice';
import './Comments.css';

const Comments = () => {
  const { postId } = useParams();
  const dispatch = useDispatch();
  const [message, setMessage] = useState('');
  const [post, setPost] = useState(null);
  const [comments, setComments] = useState(
    useSelector(selectComments).filter((comment) => comment.postId === Number(postId)),
  );
  const [nameInputValue, setNameInputValue] = useState('');
  const [emailInputValue, setEmailInputValue] = useState('');
  const [commentTextValue, setCommentTextValue] = useState('');

  useEffect(() => {
    const fetchPost = async () => {
      const result = await axios.get(`${config.API_URL.POSTS}/${postId}`);
      setPost(result ? result.data : null);
    };

    const fetchComments = async () => {
      const result = await axios.get(`${config.API_URL.COMMENTS}?postId=${postId}`);
      setComments(result ? result.data.concat(comments) : []);
    };

    fetchPost();
    fetchComments();
  }, [postId]);

  const changeName = (event) => {
    setNameInputValue(event.target.value);
  };

  const changeEmail = (event) => {
    setEmailInputValue(event.target.value);
  };

  const changeComment = (event) => {
    setCommentTextValue(event.target.value);
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    const newComment = {
      postId: Number(postId),
      id: comments[comments.length - 1] ? comments[comments.length - 1].id + 1 : 0,
      name: nameInputValue,
      email: emailInputValue,
      body: commentTextValue,
    };

    dispatch(addComment(newComment));
    setComments([...comments, newComment]);
    setNameInputValue('');
    setEmailInputValue('');
    setCommentTextValue('');
    setMessage('Comment was added successfully!');
  };

  return (
    <div className="container-fluid video-player">
      <div className="container">
        <div className="row vr-gallery">
          {post
            && (
              <div id={post.id} className="col-md-8 mb-4">
                <div className="row">
                  <div className="col-md-12 col-lg-7 pr-0 pd-md">
                    <img src="/images/gallery-img1.png" alt="" />
                  </div>
                  <div className="col-md-12 col-lg-5 light-bg cus-pd cus-arrow-left">
                    <h3>{post.title}</h3>
                    <p>{post.body}</p>
                  </div>
                </div>
              </div>
            )}
          <div id="user-comment" className="col-md-8 mb-4">
            <div className="row">
              <div className="col-md-12 col-lg-12 light-bg cus-pd cus-arrow-left">
                <h3>Leave a comment here:</h3>
                <form onSubmit={handleSubmit}>
                  <label htmlFor="name" type="control">Name:</label>
                  <input data-testid="inputName" id="name" className="input-name" name="name" type="text" value={nameInputValue} onChange={changeName} />
                  <label htmlFor="email">Email:</label>
                  <input data-testid="inputEmail" id="email" className="input-email" name="email" type="email" value={emailInputValue} onChange={changeEmail} />
                  <label htmlFor="comments">Comment:</label>
                  <textarea data-testid="inputComments" id="comments" className="comments" value={commentTextValue} onChange={changeComment} />
                  <div className="submit-container">
                    <input data-testid="buttonPublish" id="addComment" name="addComment" type="submit" value="Publish" />
                  </div>
                  { message && <p className="message">{message}</p>}
                </form>
              </div>
            </div>
          </div>
          {comments
            && (
              <div id="comments" className="col-md-8 mb-4">
                {comments.map((comment) => (
                  <div key={`comment-${comment.id}`} className="row">
                    <div className="col-md-12 col-lg-12 light-bg cus-pd cus-arrow-left">
                      <h3 className="commentName">{comment.name}</h3>
                      <h6 className="commentEmail">{comment.email}</h6>
                      <p className="commentBody">{comment.body}</p>
                    </div>
                  </div>
                ))}
              </div>
            )}
        </div>
      </div>
    </div>
  );
};

export default Comments;
