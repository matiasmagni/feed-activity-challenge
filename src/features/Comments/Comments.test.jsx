import React from 'react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { Provider } from 'react-redux';
import { fireEvent } from '@testing-library/dom';
import { act as domAct } from 'react-dom/test-utils';
import { act as testAct } from 'react-test-renderer';
import { cleanup } from '@testing-library/react';
import renderWithRouter from '../../__mocks__/renderWithRouter';
import store from '../../redux/store';
import Comments from './Comments';
import config from '../../config';

const validValues = {
  name: 'Matias Magni',
  email: 'matias.magni@gmail.com',
  comments: 'This is a test comment.',
};

const postApiData = {
  userId: 1,
  id: 1,
  title: 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
  body: 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto',
};

const commentsApiData = [
  {
    postId: 1,
    id: 1,
    name: 'id labore ex et quam laborum',
    email: 'Eliseo@gardner.biz',
    body: 'laudantium enim quasi est quidem magnam voluptate …utem quasi\nreiciendis et nam sapiente accusantium',
  },
  {
    postId: 1,
    id: 2,
    name: 'quo vero reiciendis velit similique earum',
    email: 'Jayne_Kuhic@sydney.com',
    body: 'est natus enim nihil est dolore omnis voluptatem n…iatur\nnihil sint nostrum voluptatem reiciendis et',
  },
  {
    postId: 1,
    id: 3,
    name: 'odio adipisci rerum aut animi',
    email: 'Nikita@garfield.biz',
    body: 'quia molestiae reprehenderit quasi aspernatur\naut …mus et vero voluptates excepturi deleniti ratione',
  },
];

const mockGetPost = new MockAdapter(axios);

jest.useFakeTimers();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useParams: () => ({
    postId: 1,
  }),
  useRouteMatch: () => ({ url: '/company/company-id1/team/team-id1' }),
}));

beforeEach(() => {
  jest.clearAllMocks();
  mockGetPost.reset();
});

afterEach(cleanup);

describe('Comments component tests.', () => {
  it('Comments component renders correctly.', async () => {
    let commentsComponent;

    domAct(() => {
      testAct(() => {
        commentsComponent = renderWithRouter(<Provider store={store}><Comments /></Provider>, '/comments/1');
        mockGetPost.onGet(`${config.API_URL.POSTS}/1`).reply(200, postApiData);
        mockGetPost.onGet(`${config.API_URL.COMMENTS}?postId=1`).reply(200, commentsApiData);
      });
    });

    const { asFragment } = commentsComponent;
    await expect(asFragment()).toMatchSnapshot();
  });

  it('User adds a comment after filling all fields.', () => {
    const { getByTestId, findByText } = renderWithRouter(<Provider store={store}><Comments /></Provider>, '/comments/1');
    mockGetPost.onGet(`${config.API_URL.POSTS}/1`).reply(200, postApiData);
    mockGetPost.onGet(`${config.API_URL.COMMENTS}?postId=1`).reply(200, commentsApiData);

    ['Name', 'Email', 'Comments'].forEach((elementName) => {
      fireEvent.change(getByTestId(`input${elementName}`), {
        target: {
          value: validValues[elementName.toLowerCase],
        },
      });
    });

    fireEvent.click(getByTestId('buttonPublish'));

    setTimeout(() => {
      Object.keys(validValues).forEach((i) => {
        expect(findByText(validValues[i])).toBeInTheDocument();
      });

      expect(findByText('Comment was added successfully!')).toBeVisible();
    }, 1000);

    jest.runAllTimers();
  });
});
