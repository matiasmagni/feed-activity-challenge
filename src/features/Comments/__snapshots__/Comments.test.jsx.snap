// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments component tests. Comments component renders correctly. 1`] = `
<DocumentFragment>
  <div
    class="container-fluid video-player"
  >
    <div
      class="container"
    >
      <div
        class="row vr-gallery"
      >
        <div
          class="col-md-8 mb-4"
          id="user-comment"
        >
          <div
            class="row"
          >
            <div
              class="col-md-12 col-lg-12 light-bg cus-pd cus-arrow-left"
            >
              <h3>
                Leave a comment here:
              </h3>
              <form>
                <label
                  for="name"
                  type="control"
                >
                  Name:
                </label>
                <input
                  class="input-name"
                  data-testid="inputName"
                  id="name"
                  name="name"
                  type="text"
                  value=""
                />
                <label
                  for="email"
                >
                  Email:
                </label>
                <input
                  class="input-email"
                  data-testid="inputEmail"
                  id="email"
                  name="email"
                  type="email"
                  value=""
                />
                <label
                  for="comments"
                >
                  Comment:
                </label>
                <textarea
                  class="comments"
                  data-testid="inputComments"
                  id="comments"
                />
                <div
                  class="submit-container"
                >
                  <input
                    data-testid="buttonPublish"
                    id="addComment"
                    name="addComment"
                    type="submit"
                    value="Publish"
                  />
                </div>
              </form>
            </div>
          </div>
        </div>
        <div
          class="col-md-8 mb-4"
          id="comments"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
